File: solutions/3. Longest Substring Without Repeating Characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        
        
        # alan
        sub={}
        l=0
        max_sub=0
        for i in range(len(s)):
            if s[i] in sub:
                l=max(sub[s[i]]+1,l)
            sub[s[i]]=i
            
            max_sub=max(i-l+1,max_sub)
            #print(l,i,sub,max_sub,i-l+1)
        return max_sub
                
            
        
        
        
        
        
        
        
        # neetcode
        # sub=set()
        # l,res=0,0
        # for r in range(len(s)):

File: solutions/3. Longest Substring Without Repeating Characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        
        int left=0, right=0, res=0;
        
        Set<Character> count = new HashSet<>();
        
        
        while (right<s.length()) {
            char temp = s.charAt(right);
            if (!count.contains(temp))  count.add(temp);
            else {
                while (count.contains(temp)) {
                    count.remove(s.charAt(left));
                    left++;
                }  
            }
             count.add(temp);
             res= Math.max(res, right-left+1);
             right++;
            //System.out.println(left+" "+right+" "+res+" "+count+" ");
        }
        
        return res;
        
    }
}
